### PR TITLE
Add support for -ffat-lto-objects.

### PR DESCRIPTION
**Pull request by JohanEngelen**

### CODE REVIEW

**Comment by JohanEngelen** [comment]
I think Fat LTO is still pretty experimental in LLVM, but clang provides it, so why not us. Does not impact other functionality.

File: driver/linker-gcc.cpp [hunk 1]
```diff
@@ -201,19 +201,24 @@ void ArgsBuilder::addDarwinLTOFlags() {
...
-                 (opts::linker.empty() && isLldDefaultLinker());
     addLTOGoldPluginFlags(!isLld);
   } else if (global.params.targetTriple->isOSDarwin()) {
     addDarwinLTOFlags();
   }
+
+  // Pass Fat LTO option to LLD
+  if (isLld && opts::ltoFatObjects) {
+      addLdFlag("--fat-lto-objects");
+  }
```
Comment (kinke) [inline driver/linker-gcc.cpp:221]: Hmm, why does LLD need that info? I assume the info is available in the object files themselves.

File: gen/optimizer.cpp [hunk 2]
```diff
@@ -729,7 +729,17 @@ void runOptimizationPasses(llvm::Module *M) {
   OptimizationLevel level = getOptimizationLevel();
 
   if (optLevelVal == 0) {
-    mpm = pb.buildO0DefaultPipeline(level, opts::isUsingLTO() || opts::isUsingThinLTO());
+    mpm = pb.buildO0DefaultPipeline(level, opts::isUsingLTO());
```
Comment (kinke) [inline gen/optimizer.cpp:732]: The removed thin-LTO check looks like a functional change. Not sure how much sense LTO without optimizations makes anyway though...

Comment (JohanEngelen) [reply gen/optimizer.cpp:732]: This is not a functional change. `isUsingLTO` is `true` for any type of LTO.

Comment (JohanEngelen) [reply gen/optimizer.cpp:732]: it is not needed for this PR, that's true.

File: driver/linker-gcc.cpp (hunk 1 above)
Comment (JohanEngelen) [reply driver/linker-gcc.cpp:221]: Clang also passes this. I assumed it is needed when making a library.

**Review by kinke — APPROVED** [review]

**Comment by kinke** [comment]
Wanna get this into v2.107?

**Comment by JohanEngelen** [comment]
Ah sure, we can put this in. I didn't really test it, but I don't think it hurts anyone by putting it in.

**Pull request merged**